### ISSUE TITLE
Support for setting the server level default `password_encryption` value to SCRAM-SHA-256

### ISSUE THREAD

**Issue by tspearconquest** [hashicorp/terraform-provider-azurerm#19366]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Hello,

We have created an Azure Database for Postgresql Flexible Server with the below terraform code.

```
resource "azurerm_postgresql_flexible_server" "default" {
  name                = var.server_name
  resource_group_name = var.rg_name
  location            = var.region

  administrator_login    = var.admin_user
  administrator_password = var.admin_password

  backup_retention_days = var.backup_retention_days
  #checkov:skip=CKV_AZURE_136:Geo redundant backups are expensive so we are opting out but leaving this as a future option
  geo_redundant_backup_enabled = var.geo_redundant_backups

  delegated_subnet_id = var.dedicated_subnet_id
  private_dns_zone_id = var.private_dns_zone_id

  sku_name   = var.sku_name
  storage_mb = var.storage_mb
  version    = var.server_version
  zone       = var.default_availability_zone

  dynamic "high_availability" {
    for_each = var.high_availability == true ? ["ha"] : []

    content {
      mode                      = "ZoneRedundant"
      standby_availability_zone = var.standby_availability_zone
    }
  }

  maintenance_window {
    day_of_week = var.maintenance_start_day
  }

  tags = merge(local.common_tags)
  lifecycle {
    ignore_changes = [
      high_availability[0].standby_availability_zone,
      tags["createdDate"],
      zone
    ]
  }
}
```

Upon creating the server and logging in with the admin user, we found that Azure has set the default value for the `password_encryption` server parameter to `MD5`, however as per the [Postgres 14 documentation](https://www.postgresql.org/docs/14/runtime-config-connection.html#GUC-PASSWORD-ENCRYPTION): The default value for `password_encryption` is supposed to be `SCRAM-SHA-256`.

This inconsistency does not appear in the Azure documentation and while we could change the value with the below blocks of Terraform code, there are side effects related to doing so, which I will detail below:

```
resource "azurerm_postgresql_flexible_server_configuration" "azure_accepted_password_auth_method" {
  name      = "azure.accepted_password_auth_method"
  server_id = azurerm_postgresql_flexible_server.default.id
  value     = "SCRAM-SHA-256"
}

resource "azurerm_postgresql_flexible_server_configuration" "password_encryption" {
  name      = "password_encryption"
  server_id = azurerm_postgresql_flexible_server.default.id
  value     = "SCRAM-SHA-256"
}
```

As mentioned there are several side effects of using these:

1. The administrator password upon initial creation is set to MD5. So changing the accepted password auth method to only `SCRAM-SHA-256` prevents anyone from logging in with that user in order to make additional changes (either with Terraform via the `postgres` provider, or with the postgres CLI.
2. If we change the `azure.accepted_password_auth_method` value to allow both `MD5` and `SCRAM-SHA-256`, we are able to login as the administrator user, and then we can manually reset the administrator password to the same one, which saves the password's `authtype` field properly to `SCRAM-256` as below, however that then puts the terraform state out of sync with the server configuration which causes Terraform to reset the password again back to MD5:
    ```
    postgres=> SELECT * from azure_roles_authtype() where rolename = 'myadminuser';
        rolename    | authtype
    ----------------+-----------
     myadminuser | SCRAM-256
    (1 row)
    ```
3. According to the Postgres 14 [docs](https://www.postgresql.org/docs/14/runtime-config-connection.html#GUC-PASSWORD-ENCRYPTION), the default password encryption method is `SCRAM-SHA-256`, however since Azure defaults to `MD5`, we are unable to properly secure the administrator account due to the above mentioned side effects:

> password_encryption (enum)
> When a password is specified in [CREATE ROLE](https://www.postgresql.org/docs/14/sql-createrole.html) or [ALTER ROLE](https://www.postgresql.org/docs/14/sql-alterrole.html), this parameter determines the algorithm to use to encrypt the password. Possible values are scram-sha-256, which will encrypt the password with SCRAM-SHA-256, and md5, which stores the password as an MD5 hash. **_The default is scram-sha-256._**

4. Additionally, the Azure Portal password reset function also defaults to `MD5` even if `azure.accepted_password_auth_method` is set to only `SCRAM-SHA-256`, which then again breaks logins. As mentioned at the bottom of this issue, we have a support case open and it has been escalated to Engineering, as the portal should default to `SCRAM-SHA-256` if it is enabled, because as per the above Postgres docs, `SCRAM-SHA-256` is the default.

This request is to add an additional attribute to the `azurerm_postgresql_flexible_server` resource which allows us to set the default for both of the above server parameters to `SCRAM-SHA-256` at server creation time (rather than doing so with an `azurerm_postgresql_flexible_server_configuration` block) so that new servers are automatically created with only SCRAM authentication and password encryption, in order to avoid the issues outlined.

### New or Affected Resource(s)/Data Source(s)

azurerm_postgresql_flexible_server

### Potential Terraform Configuration

```hcl
resource "azurerm_postgresql_flexible_server" "default" {
  name                = var.server_name
  resource_group_name = var.rg_name
  location            = var.region

  administrator_login    = var.admin_user
  administrator_password = var.admin_password

  backup_retention_days = var.backup_retention_days
  #checkov:skip=CKV_AZURE_136:Geo redundant backups are expensive so we are opting out but leaving this as a future option
  geo_redundant_backup_enabled = var.geo_redundant_backups

  delegated_subnet_id = var.dedicated_subnet_id
  private_dns_zone_id = var.private_dns_zone_id

  sku_name   = var.sku_name
  storage_mb = var.storage_mb
  version    = var.server_version
  zone       = var.default_availability_zone

  password_encryption            = "SCRAM-SHA-256"
  password_authentication_method = ["SCRAM-SHA-256"]

  dynamic "high_availability" {
    for_each = var.high_availability == true ? ["ha"] : []

    content {
      mode                      = "ZoneRedundant"
      standby_availability_zone = var.standby_availability_zone
    }
  }

  maintenance_window {
    day_of_week = var.maintenance_start_day
  }

  tags = merge(local.common_tags)
  lifecycle {
    ignore_changes = [
      high_availability[0].standby_availability_zone,
      tags["createdDate"],
      zone
    ]
  }
}
```

### References

We have a support ticket open with Azure Support referencing a Portal bug around the Password Reset feature related to this, and they have acknowledged that there is a bug and have escalated the issue to engineering. I'm not sure of the best way to provide the ticket number, but would prefer not to put it in this issue.

**Comment by neil-yechenwei** [comment]
Thanks for raising this issue. Seems it's a service API issue. As service API doesn't provide the ability to directly set the password encryption at creation time, so we cannot support it in azurerm_postgresql_flexible_server.

**Comment by tspearconquest** [comment]
Hello,

Understood. I have a support case open. Is there some way I can securely transmit the case number to you for escalation with the service API team?

This is a regulatory requirement we need to comply with, and we are a Microsoft Partner, so it would be great if we can get your help in raising the case internally.

Thanks

**Comment by ChristianOelsner** [comment]
Are there any news regarding this?
We are also facing this challenge.

**Comment by gcbikram** [comment]
@neil-yechenwei are there any updates on this? we also have a support ticket outstanding.

**Comment by gbowerman** [comment]
FYI the defaults for Azure PostgreSQL Flexible Server v14 and higher are changing this week. 
For new PG14 and higher server deployments the password_encryption server parameter will now default to "SCRAM-SHA-256".
The azure.accepted_password_auth_method parameter default being modified to be "MD5,SCRAM-SHA-256".

**Comment by tspearconquest** [comment]
Fantastic news! Will this apply to only new infrastructure created after the change?

Is there any steps we would need to take to be able to adopt this change on pre-existing Azure PostgreSQL Flexible Server v14 servers we are already using in production?

**Comment by gbowerman** [comment]
> Fantastic news! Will this apply to only new infrastructure created after the change?

Yes, the current change in default encryption and auth parameter settings that's rolling out will only be for new PostgreSQL 14 and higher database server deployments. Existing servers will keep the config parameters they currently have to minimize the possibility of breaking change.

> Is there any steps we would need to take to be able to adopt this change on pre-existing Azure PostgreSQL Flexible Server v14 servers we are already using in production?

It is an ongoing recommended practice to manually set encryption and auth parameter settings to SCRAM-SHA-265 for existing servers if it's not already been done.

**Comment by tspearconquest** [comment]
Thank you again. And this is happening on the API side, so we won't see this is the release notes for the next azurerm terraform provider release, most likely?

**Comment by gbowerman** [comment]
> Thank you again. And this is happening on the API side, so we won't see this in the release notes for the next azurerm terraform provider release, most likely?

Yes, you'll see the updates in the PostgreSQL Flexible Server July release notes (https://learn.microsoft.com/en-us/azure/postgresql/flexible-server/release-notes#release-july-2024) and in the parameter docs for 14+, like here for example: https://learn.microsoft.com/en-us/azure/postgresql/flexible-server/server-parameters-table-customized-options?pivots=postgresql-14#azureaccepted_password_auth_method

Happy to work with the azurerm terraform provider owner if you anyone has a contact?

**Comment by tspearconquest** [comment]
Looking over the postgres flex docs for tf, it doesn't seem any changes are needed on the client side other than making use of the new version of the API.

CC: @tombuildsstuff is the only contact I know.

**Comment by gbowerman** [comment]
> Looking over the postgres flex docs for tf, it doesn't seem any changes are needed on the client side other than making use of the new version of the API.

You shouldn't need to use a new API either. Deploy v14 or later server and you'll see the new defaults. (Change is rolling out worldwide this week, at the time of writing it's already in eastasia, canadacentral, westcentralus).

**Comment by tspearconquest** [comment]
No longer an issue; simply forgot to close this until now.
